src/main/scala/Validator.scala

```diff
@@ -1,115 +1,387 @@
+    var errors: Array[ErrorWithCsvContext] = Array()
+
+    val parserAfterSkippedRows = parser.asScala.drop(dialect.skipRows)
     val table = schema.tables(tableUri.toString)
+    val childTableForeignKeys =
+      Map[ChildTableForeignKey, mutable.Set[KeyWithContext]]()
+    val parentTableForeignKeyReferences =
+      Map[ParentTableForeignKeyReference, mutable.Set[KeyWithContext]]()
+    val allPrimaryKeyValues: mutable.Set[List[Any]] =
+      Set() // List of type Any with same values are not added again in a Set, whereas Array of Type any behaves
+    // differently.
 
-    // List of type Any with same values are not added again in a Set, whereas Array of Type any behaves differently.
-    var allPrimaryKeyValues: Set[List[Any]] = Set()
-    for (row <- parser.asScala) {
-      if (row.getRecordNumber == 1 && csvHeaderExpected) {
-        validateHeader(schema, row, tableUri)
+    parserAfterSkippedRows
+      .map(row =>
+        (
+          row,
+          parseRow(
+            schema,
+            tableUri,
+            dialect,
+            table,
+            row
+          )
+        )
+      )
+      .foreach {
+        case (row, validateRowOutput) => {
+          errors ++= validateRowOutput.warningsAndErrors.errors
+          warnings ++= validateRowOutput.warningsAndErrors.warnings
+          setChildTableForeignKeys(
+            validateRowOutput,
+            childTableForeignKeys
+          )
+          setParentTableForeignKeyReferences(
+            validateRowOutput,
+            parentTableForeignKeyReferences
+          )
+          validatePrimaryKey(allPrimaryKeyValues, row, validateRowOutput)
+            .ifDefined(e => {
+              errors :+= e
+            })
+        }
+      }
+
+    (
+      WarningsAndErrors(warnings, errors),
+      childTableForeignKeys,
+      parentTableForeignKeyReferences
+    )
+  }
+
+  private def parseRow(
+      schema: TableGroup,
+      tableUri: URI,
+      dialect: Dialect,
+      table: Table,
+      row: CSVRecord
+  ): ValidateRowOutput = {
+    if (row.getRecordNumber == 1 && dialect.header) {
+      val warningsAndErrors = schema.validateHeader(row, tableUri.toString)
+      ValidateRowOutput(warningsAndErrors)
+    } else {
+      if (row.size == 0) {
+        val blankRowError = ErrorWithCsvContext(
+          "Blank rows",
+          "structure",
+          row.getRecordNumber.toString,
+          "",
+          "",
+          ""
+        )
+        val warningsAndErrors =
+          WarningsAndErrors(errors = Array(blankRowError))
+        ValidateRowOutput(warningsAndErrors)
       } else {
-        if (row.size == 0) {
-          warnings :+= ErrorWithCsvContext(
-            "Blank rows",
-            "structure",
-            row.getRecordNumber.toString,
-            "",
-            "",
-            ""
+        if (table.columns.length >= row.size()) { // todo: Ensure there's a error defined somewhere else when this is not true.
```

Comment:
Checked this todo?

Reply:
I have added an error here for now. The ruby equivalent of this only adds this error when ` if !row.empty? && row.size != @expected_columns` where expected columns does not make much sense.  

I think our approach is better here.